torchmetrics/retrieval/recall_precision.py

```diff
@@ -0,0 +1,297 @@
+            - ``'error'``: raise a ``ValueError``
+
+        ignore_index:
+            Ignore predictions where the target is equal to this number.
+        compute_on_step:
+            Forward only calls ``update()`` and returns None if this is set to False.
+
+            .. deprecated:: v0.8
+                Argument has no use anymore and will be removed v0.9.
+
+        kwargs:
+            Additional keyword arguments, see :ref:`Metric kwargs` for more info.
+
+    Raises:
+        ValueError:
+            If ``empty_target_action`` is not one of ``error``, ``skip``, ``neg`` or ``pos``.
+        ValueError:
+            If ``ignore_index`` is not `None` or an integer.
+        ValueError:
+            If ``max_k`` parameter is not `None` or an integer larger than 0.
+
+    Example:
+        >>> from torchmetrics import RetrievalPrecisionRecallCurve
+        >>> indexes = tensor([0, 0, 0, 0, 1, 1, 1])
+        >>> preds = tensor([0.4, 0.01, 0.5, 0.6, 0.2, 0.3, 0.5])
+        >>> target = tensor([True, False, False, True, True, False, True])
+        >>> r = RetrievalPrecisionRecallCurve(max_k=4)
+        >>> precisions, recalls, top_k = r(preds, target, indexes=indexes)
+        >>> precisions
+        tensor([1.0000, 0.5000, 0.6667, 0.5000])
+        >>> recalls
+        tensor([0.5000, 0.5000, 1.0000, 1.0000])
+        >>> top_k
+        tensor([1, 2, 3, 4])
+    """
+
+    higher_is_better = True
+
+    def __init__(
+        self,
+        max_k: Optional[int] = None,
+        adaptive_k: bool = False,
+        empty_target_action: str = "neg",
+        ignore_index: Optional[int] = None,
+        compute_on_step: Optional[bool] = None,
+        **kwargs: Dict[str, Any],
+    ) -> None:
+        super().__init__(compute_on_step=compute_on_step, **kwargs)
+        self.allow_non_binary_target = False
+
+        empty_target_action_options = ("error", "skip", "neg", "pos")
+        if empty_target_action not in empty_target_action_options:
+            raise ValueError(f"Argument `empty_target_action` received a wrong value `{empty_target_action}`.")
+
+        self.empty_target_action = empty_target_action
+
+        if ignore_index is not None and not isinstance(ignore_index, int):
+            raise ValueError("Argument `ignore_index` must be an integer or None.")
+
+        self.ignore_index = ignore_index
+
+        self.add_state("indexes", default=[], dist_reduce_fx=None)
+        self.add_state("preds", default=[], dist_reduce_fx=None)
+        self.add_state("target", default=[], dist_reduce_fx=None)
+
+        if (max_k is not None) and not (isinstance(max_k, int) and max_k > 0):
+            raise ValueError("`max_k` has to be a positive integer or None")
+
+        if not isinstance(adaptive_k, bool):
+            raise ValueError("`adaptive_k` has to be a boolean")
+
+        self.max_k = max_k
+        self.adaptive_k = adaptive_k
```

Comment:
nitpicking: usually we initialize the states as the last part of `__init__`:
```suggestion
        if (max_k is not None) and not (isinstance(max_k, int) and max_k > 0):
            raise ValueError("`max_k` has to be a positive integer or None")
        self.max_k = max_k
        
        if not isinstance(adaptive_k, bool):
            raise ValueError("`adaptive_k` has to be a boolean")
        self.adaptive_k = adaptive_k
        
        self.add_state("indexes", default=[], dist_reduce_fx=None)
        self.add_state("preds", default=[], dist_reduce_fx=None)
        self.add_state("target", default=[], dist_reduce_fx=None)
```